glib2-devel: Update to 2.62.2

Lord-Kamina opened the pull request:
I’ve updated this for meson; refreshed patch offsets and done my best to duplicate the x11/quartz selection logic. It builds but has not been THOROUGHLY tested. As for tests, about 12 of them fail; a few of those were failing in 2.58.3 too; not sure if the new failures are my fault, new bugs, new tests or what.

P.S. @dbevans you can have your tarballs and eat them too. :P

#### Description

<!-- Note: it is best to make pull requests from a branch rather than from master -->

###### Type(s)
<!-- update (title contains ": U(u)pdate to"), submission (new Portfile) and CVE Identifiers are auto-detected, replace [ ] with [x] to select -->

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on
<!-- Generate version information with this command in shell:
    echo "macOS $(sw_vers -productVersion) $(sw_vers -buildVersion)"; echo "Xcode $(xcodebuild -version | awk '{print $NF}' | tr '\n' ' ')"
-->
macOS 10.14.6
Xcode 11.1

###### Verification <!-- (delete not applicable items) -->
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] checked your Portfile with `port lint`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

<!-- Use "skip notification" (surrounded with []) to avoid notifying maintainers -->

ryandesign commented:
Thanks for attempting this update. As you may know, getting this right is going to be very difficult, and also very important since this is a fundamental dependency of so many other ports.

The commit message should reference the ticket about this update: https://trac.macports.org/ticket/58261

Let's please update glib2-devel first so that we can verify it builds and works on all the OS versions before applying the same changes to glib2.

I do not want the libintl.patch and universal.patch patchfiles' names changed. The `port lint` advice about patchfile naming is outdated and does not represent our current practices. The `port lint` message should be changed in MacPorts base. https://trac.macports.org/ticket/59695

Hardcoding `${workpath}/build` is not a good idea. That is an implementation detail of the meson portgroup that portfiles should not rely on. Use `${build.dir}` instead.

I see you've patched meson.build to set glib_memory_barrier_needed to true if host_machine.endian() is "big" and false otherwise. According to meson documentation, host_machine refers to the machine for which we're building, but when building universal, we would build for multiple host machines. The only way meson could understand that properly through a single variable is if we hide from meson the fact that we are building universal, e.g. by building for each architecture separately, i.e. by using the muniversal portgroup. Using the muniversal portgroup can cause a bunch of problems, which is why we stopped using it in the glib2 portfile in ca3bacb8c5e7f35c5f71a9b3ca811691aa51836a. If meson gives us no option but to go back to using it, then we need to consider whether this will reintroduce the problems that no longer using the portgroup fixed; see https://trac.macports.org/ticket/20372.

I see that glib2 has the clang_dependency portgroup but that meson doesn't. Does adding the meson portgroup/dependency to glib2 cause problems in that regard, i.e. does it mean that we will need to add the clang_dependency portgroup to meson and all of its dependencies recursively?

The build failure reported in the CI logs above needs to be fixed:

```
meson.build:2003:0: ERROR: <ExternalProgram 'python' -> ['/usr/bin/python']> is not a valid python or it is missing setuptools
```

Lord-Kamina commented:
> Thanks for attempting this update. As you may know, getting this right is going to be very difficult, and also very important since this is a fundamental dependency of so many other ports.
> 
> The commit message should reference the ticket about this update: https://trac.macports.org/ticket/58261

Ran a search here and also looked into whether anybody had a fork with a partial update but didn't really look in trac.

> 
> Let's please update glib2-devel first so that we can verify it builds and works on all the OS versions before applying the same changes to glib2.

Sure.

> I do not want the libintl.patch and universal.patch patchfiles' names changed. The `port lint` advice about patchfile naming is outdated and does not represent our current practices. The `port lint` message should be changed in MacPorts base. https://trac.macports.org/ticket/59695
>

Ok; in fact, that was a last-minute change right before I pushed the PR.

> Hardcoding `${workpath}/build` is not a good idea. That is an implementation detail of the meson portgroup that portfiles should not rely on. Use `${build.dir}` instead.
> 

I looked for such a variable in the portfile reference and didn't find it, which is why I hardcoded it.

> I see you've patched meson.build to set glib_memory_barrier_needed to true if host_machine.endian() is "big" and false otherwise. According to meson documentation, host_machine refers to the machine for which we're building, but when building universal, we would build for multiple host machines. The only way meson could understand that properly through a single variable is if we hide from meson the fact that we are building universal, e.g. by building for each architecture separately, i.e. by using the muniversal portgroup. Using the muniversal portgroup can cause a bunch of problems, which is why we stopped using it in the glib2 portfile in [ca3bacb](https://github.com/macports/macports-ports/commit/ca3bacb8c5e7f35c5f71a9b3ca811691aa51836a). If meson gives us no option but to go back to using it, then we need to consider whether this will reintroduce the problems that no longer using the portgroup fixed; see https://trac.macports.org/ticket/20372.
> 

As I said, I merely replicated the logic of what was there before, at least what I thought/understood it to be doing. So, completely open to suggestions and corrections there.

> I see that glib2 has the clang_dependency portgroup but that meson doesn't. Does adding the meson portgroup/dependency to glib2 cause problems in that regard, i.e. does it mean that we will need to add the clang_dependency portgroup to meson and all of its dependencies recursively?
> 

Not that I've noticed; maybe somebody can weigh in further?

> The build failure reported in the CI logs above needs to be fixed:
> 
> ```
> meson.build:2003:0: ERROR: <ExternalProgram 'python' -> ['/usr/bin/python']> is not a valid python or it is missing setuptools
> ```

Oh, I didn't notice that. Only saw the `Bash exited with code '1'.` error.

Anyway, will try to address these later today.

Lord-Kamina commented:
Is universal only i386/x86_64 or are we still concerned about ppc?

ryandesign commented:
> > I see you've patched meson.build to set glib_memory_barrier_needed to true if host_machine.endian() is "big" and false otherwise.
> 
> As I said, I merely replicated the logic of what was there before, at least what I thought/understood it to be doing.

But that's not the same. The old universal.patch modifies the configure script:

```
 	if test x"$g_memory_barrier_needed" != xno; then
 	  echo >>$outfile
+	  echo "#ifdef __BIG_ENDIAN__" >>$outfile
 	  echo "#define G_ATOMIC_OP_MEMORY_BARRIER_NEEDED 1" >>$outfile
+	  echo "#endif" >>$outfile
 	fi
```

The intention was that this was *writing out a file* which would contain:

```
#ifdef __BIG_ENDIAN__
#define G_ATOMIC_OP_MEMORY_BARRIER_NEEDED 1
#endif
```

This file would be processed by the compiler at build time. So `G_ATOMIC_OP_MEMORY_BARRIER_NEEDED` would get defined on PowerPC parts of the build and not on Intel parts of the build, even if we're compiling them at the same time using multiple `-arch` flags.

However, there may be a bug here, in that this was only going into the file if `test x"$g_memory_barrier_needed" != xno`. I didn't look into how `$g_memory_barrier_needed` gets set, but presumably it gets set on PowerPC systems only, whereas we would always want to write this to the file. So this may be a bug in the existing port, with the effect that a PowerPC/Intel universal build is only correct for PowerPC systems if performed on a PowerPC build machine.

But the point is that the decision about whether to use the atomic memory barrier should happen at build time, not at configure time as you had proposed.

But now it looks like you've force-pushed and are no longer making any attempts to set the memory barrier needed correctly.

I see that there is:

```
#mesondefine G_ATOMIC_OP_MEMORY_BARRIER_NEEDED
```

I don't know what `#mesondefine` is but maybe this can still be enclosed in an `#ifdef __BIG_ENDIAN__` somehow.

> > does it mean that we will need to add the clang_dependency portgroup to meson and all of its dependencies recursively?
> 
> Not that I've noticed;

How did you check? Do you know what the portgroup is for? You would not notice a problem unless you are on one of the systems for which the portgroup exists (OS X 10.8 and earlier with libc++) and have not yet installed the compiler dependencies. The purpose of the portgroup is to prevent a circular dependency situation from existing.

> Evidently I should prefer a python provided by macports; are there any details on how should I decide _which_ to use? Can I rely on `${prefix}/bin/python3` existing, for example?

No, you cannot rely on `${prefix}/bin/python3` existing. Usually I would recommend using the latest stable version of python 3. Currently, that's 3.8. Ideally, base everything on a single variable in the Portfile that can be changed from 3.8 to a different version in the future.

However again, that would require adding the clang_dependency portgroup to python38 and all of its recursive dependencies. To avoid that, use python 3.7 instead, since python37 already uses that portgroup, as presumably do its dependencies.

> Is universal only i386/x86_64 or are we still concerned about ppc?

We should care about both Intel and PowerPC systems. There are still people using MacPorts on PowerPC Macs and glib2 is a dependency of a lot of other software. Breaking glib2 on PowerPC would break the glib2-using ports that currently work there.

Lord-Kamina commented:
> > > I see you've patched meson.build to set glib_memory_barrier_needed to true if host_machine.endian() is "big" and false otherwise.
> > 
> > 
> > As I said, I merely replicated the logic of what was there before, at least what I thought/understood it to be doing.
> 
> But that's not the same. The old universal.patch modifies the configure script:
> 
> ```
>  	if test x"$g_memory_barrier_needed" != xno; then
>  	  echo >>$outfile
> +	  echo "#ifdef __BIG_ENDIAN__" >>$outfile
>  	  echo "#define G_ATOMIC_OP_MEMORY_BARRIER_NEEDED 1" >>$outfile
> +	  echo "#endif" >>$outfile
>  	fi
> ```
> 
> The intention was that this was _writing out a file_ which would contain:
> 
> ```
> #ifdef __BIG_ENDIAN__
> #define G_ATOMIC_OP_MEMORY_BARRIER_NEEDED 1
> #endif
> ```
> 
> This file would be processed by the compiler at build time. So `G_ATOMIC_OP_MEMORY_BARRIER_NEEDED` would get defined on PowerPC parts of the build and not on Intel parts of the build, even if we're compiling them at the same time using multiple `-arch` flags.
> 
> However, there may be a bug here, in that this was only going into the file if `test x"$g_memory_barrier_needed" != xno`. I didn't look into how `$g_memory_barrier_needed` gets set, but presumably it gets set on PowerPC systems only, whereas we would always want to write this to the file. So this may be a bug in the existing port, with the effect that a PowerPC/Intel universal build is only correct for PowerPC systems if performed on a PowerPC build machine.
> 
> But the point is that the decision about whether to use the atomic memory barrier should happen at build time, not at configure time as you had proposed.
> 
> But now it looks like you've force-pushed and are no longer making any attempts to set the memory barrier needed correctly.
> 
> I see that there is:
> 
> ```
> #mesondefine G_ATOMIC_OP_MEMORY_BARRIER_NEEDED
> ```
> 
> I don't know what `#mesondefine` is but maybe this can still be enclosed in an `#ifdef __BIG_ENDIAN__` somehow.
> 

I think the vanilla meson.build essentially would do the same thing the old patch was doing then (with possible the save caveat you mentioned), which is why I removed my patch. `#mesondefine` behaves differently depending on the value of its argument. If `true` it expands to `#define`; if `false` it expands to `#undef` and if it's a number it defines the macro with that value.

> > > does it mean that we will need to add the clang_dependency portgroup to meson and all of its dependencies recursively?
> > 
> > 
> > Not that I've noticed;
> 
> How did you check? Do you know what the portgroup is for? You would not notice a problem unless you are on one of the systems for which the portgroup exists (OS X 10.8 and earlier with libc++) and have not yet installed the compiler dependencies. The purpose of the portgroup is to prevent a circular dependency situation from existing.
> 

In that case, I have no way of checking this and we'll need somebody else to help test it. I cannot even build i386 with Mojave and Xcode11, so...

> > Evidently I should prefer a python provided by macports; are there any details on how should I decide _which_ to use? Can I rely on `${prefix}/bin/python3` existing, for example?
> 
> No, you cannot rely on `${prefix}/bin/python3` existing. Usually I would recommend using the latest stable version of python 3. Currently, that's 3.8. Ideally, base everything on a single variable in the Portfile that can be changed from 3.8 to a different version in the future.
> 
> However again, that would require adding the clang_dependency portgroup to python38 and all of its recursive dependencies. To avoid that, use python 3.7 instead, since python37 already uses that portgroup, as presumably do its dependencies.
> 

I noticed anyway that `configure.python` doesn't work for the purposes of this port; I added `python37` as a lib dependency and decided to set `configure.python` using a dirty hack to get whatever version of python meson depends on (currently 3.7); I would have loved to have been able to declare the dependency using the same trick but it appears this is not supported.
Either way, this doesn't work right away because of the way glib2 and meson look for python. Glib2's top `meson.build` is set to look for "python3" which won't work unless `port select` was run beforehand; it uses "python" as a backup (which is why the CI was defaulting to system python). Reading online, I realized meson itself can fallback to whatever python it is itself using if the argument to `find_installation` is empty, so I used `reinplace` in the post-patch phase to fix that (haven't committed this yet, which is why I'm explaining it)

> > Is universal only i386/x86_64 or are we still concerned about ppc?
> 
> We should care about both Intel and PowerPC systems. There are still people using MacPorts on PowerPC Macs and glib2 is a dependency of a lot of other software. Breaking glib2 on PowerPC would break the glib2-using ports that currently work there.

So, same as for clang_dependency, we're gonna need somebody to help us test this.

ryandesign commented:
> I think the vanilla meson.build essentially would do the same thing the old patch was doing then (with possible the save caveat you mentioned), which is why I removed my patch. `#mesondefine` behaves differently depending on the value of its argument. If `true` it expands to `#define`; if `false` it expands to `#undef` and if it's a number it defines the macro with that value.

And how is that truth or falseness set by meson? With this block of code in meson.build:

```
# We need a more robust approach here...
host_cpu_family = host_machine.cpu_family()
if host_cpu_family == 'x86' or host_cpu_family == 'x86_64' or host_cpu_family == 's390' or host_cpu_family == 's390x' or host_cpu_family.startswith('arm') or host_cpu_family == 'aarch64' or host_cpu_family.startswith('crisv32') or host_cpu_family.startswith('etrax')
  glib_memory_barrier_needed = false
elif host_cpu_family.startswith('sparc') or host_cpu_family.startswith('alpha') or host_cpu_family.startswith('powerpc') or host_cpu_family == 'ia64'
  glib_memory_barrier_needed = true
else
  warning('Unknown host cpu: ' + host_cpu_family)
  glib_memory_barrier_needed = true
endif
glibconfig_conf.set('G_ATOMIC_OP_MEMORY_BARRIER_NEEDED', glib_memory_barrier_needed)
```

That is evaluated once, at the time meson is run. It will only be correct for one arch. It needs to be done by the compiler at build time, so that it can be correct for each arch as it's built.

I can fire up a PowerBook to test this on 10.4 and 10.5 once we have something that looks like it might work.

> configure.python

You're right, dependencies have to be determinable at portindex time. They can't be based on the results of files installed by other ports.

I don't think I would bother trying to be tricky to determine which python meson is using. We need to declare a specific dependency, on a a python that uses the clang_dependency portgroup, so we need to use that python, even if meson is using a different one. If the glib2 build is looking for "python3" or "python" that won't do and will have to be changed.

Lord-Kamina commented:
> > I think the vanilla meson.build essentially would do the same thing the old patch was doing then (with possible the save caveat you mentioned), which is why I removed my patch. `#mesondefine` behaves differently depending on the value of its argument. If `true` it expands to `#define`; if `false` it expands to `#undef` and if it's a number it defines the macro with that value.
> 
> And how is that truth or falseness set by meson? With this block of code in meson.build:
> 
> ```
> # We need a more robust approach here...
> host_cpu_family = host_machine.cpu_family()
> if host_cpu_family == 'x86' or host_cpu_family == 'x86_64' or host_cpu_family == 's390' or host_cpu_family == 's390x' or host_cpu_family.startswith('arm') or host_cpu_family == 'aarch64' or host_cpu_family.startswith('crisv32') or host_cpu_family.startswith('etrax')
>   glib_memory_barrier_needed = false
> elif host_cpu_family.startswith('sparc') or host_cpu_family.startswith('alpha') or host_cpu_family.startswith('powerpc') or host_cpu_family == 'ia64'
>   glib_memory_barrier_needed = true
> else
>   warning('Unknown host cpu: ' + host_cpu_family)
>   glib_memory_barrier_needed = true
> endif
> glibconfig_conf.set('G_ATOMIC_OP_MEMORY_BARRIER_NEEDED', glib_memory_barrier_needed)
> ```
> 
> That is evaluated once, at the time meson is run. It will only be correct for one arch. It needs to be done by the compiler at build time, so that it can be correct for each arch as it's built.
> 
> I can fire up a PowerBook to test this on 10.4 and 10.5 once we have something that looks like it might work.
> 

In this case, we should change the config.h to include a similar check.

> > configure.python
> 
> You're right, dependencies have to be determinable at portindex time. They can't be based on the results of files installed by other ports.
> 
> I don't think I would bother trying to be tricky to determine which python meson is using. We need to declare a specific dependency, on a a python that uses the clang_dependency portgroup, so we need to use that python, even if meson is using a different one. If the glib2 build is looking for "python3" or "python" that won't do and will have to be changed.

Which is why I went with python37 but I set the configure.python using a shell script to determine what meson uses by passing the output of `port deps meson` though grep, cut and sed. My point is, since meson can use its own python as a fallback, it would have been ideal to ensure the dependency would always have been that same one but in the end it's just two more numbers maintainers will have to change down the road. Not that big a deal.

Lord-Kamina commented:
How about this, @ryandesign ?

Lord-Kamina commented:
P.S. I realize the whole programmatic python version approach is probably doomed but I want to see if I can get it working properly; else I'll just use a hardcoded version *shrugs*

Lord-Kamina commented:
I have noticed there's apparently something wrong with `glib-mkenums`
Trying to build glade results in `glib-mkenums`spawning several `env` processes that sit around apparently doing nothing.

dbevans commented:
@Lord-Kamina: suggest the following patch to keep from hard-coding the python path in your meson.build patch file, keeping the path selection strictly in Portfile.

```
diff --git a/devel/glib2-devel/Portfile b/devel/glib2-devel/Portfile
index 49e4740a170..8e59e040346 100644
--- a/devel/glib2-devel/Portfile
+++ b/devel/glib2-devel/Portfile
@@ -81,7 +81,8 @@ configure.python            ${prefix}/bin/python3.7
 configure.env-append        PERL_PATH=${configure.perl}
 
 post-patch {
-    reinplace -W ${worksrcpath} "s|@PREFIX@|${prefix}|g" gio/gdbusprivate.c gio/xdgmime/xdgmime.c glib/gi18n-lib.h glib/gi18n.h gio/gdbusaddress.c meson.build
+    reinplace -W ${worksrcpath} "s|@PYTHON@|${configure.python}|" meson.build
+    reinplace -W ${worksrcpath} "s|@PREFIX@|${prefix}|g" gio/gdbusprivate.c gio/xdgmime/xdgmime.c glib/gi18n-lib.h glib/gi18n.h gio/gdbusaddress.c
     reinplace "s|data_dirs = \"/usr|data_dirs = \"${prefix}/share:/usr|g" ${worksrcpath}/glib/gutils.c
     reinplace "s|path = \"/bin|path = \"${prefix}/bin:/bin|g" ${worksrcpath}/glib/gutils.c ${worksrcpath}/glib/gspawn.c    
 }
diff --git a/devel/glib2-devel/files/patch-meson-build-python-path.diff b/devel/glib2-devel/files/patch-meson-build-python-path.diff
index c400808085d..bb8b8a3ef60 100644
--- a/devel/glib2-devel/files/patch-meson-build-python-path.diff
+++ b/devel/glib2-devel/files/patch-meson-build-python-path.diff
@@ -8,11 +8,11 @@ GLib2 tries to find "python3" and if it can't find it, it will go for "python";
  glib_conf.set('HAVE_PROC_SELF_CMDLINE', have_proc_self_cmdline)
  
 -python = import('python').find_installation('python3')
-+python = import('python').find_installation('')
++python = import('python').find_installation('@PYTHON@')
  # used for '#!/usr/bin/env <name>'
 -python_name = 'python3'
-+python_name = '@PREFIX@/bin/python3.7'
++python_name = '@PYTHON@'
  
  python_version = python.language_version()
  python_version_req = '>=3.4'
- 
\ No newline at end of file
+
```

Lord-Kamina commented:
> @Lord-Kamina: suggest the following patch to keep from hard-coding the python path in your meson.build patch file, keeping the path selection strictly in Portfile.
> 
> ```
> diff --git a/devel/glib2-devel/Portfile b/devel/glib2-devel/Portfile
> index 49e4740a170..8e59e040346 100644
> --- a/devel/glib2-devel/Portfile
> +++ b/devel/glib2-devel/Portfile
> @@ -81,7 +81,8 @@ configure.python            ${prefix}/bin/python3.7
>  configure.env-append        PERL_PATH=${configure.perl}
>  
>  post-patch {
> -    reinplace -W ${worksrcpath} "s|@PREFIX@|${prefix}|g" gio/gdbusprivate.c gio/xdgmime/xdgmime.c glib/gi18n-lib.h glib/gi18n.h gio/gdbusaddress.c meson.build
> +    reinplace -W ${worksrcpath} "s|@PYTHON@|${configure.python}|" meson.build
> +    reinplace -W ${worksrcpath} "s|@PREFIX@|${prefix}|g" gio/gdbusprivate.c gio/xdgmime/xdgmime.c glib/gi18n-lib.h glib/gi18n.h gio/gdbusaddress.c
>      reinplace "s|data_dirs = \"/usr|data_dirs = \"${prefix}/share:/usr|g" ${worksrcpath}/glib/gutils.c
>      reinplace "s|path = \"/bin|path = \"${prefix}/bin:/bin|g" ${worksrcpath}/glib/gutils.c ${worksrcpath}/glib/gspawn.c    
>  }
> diff --git a/devel/glib2-devel/files/patch-meson-build-python-path.diff b/devel/glib2-devel/files/patch-meson-build-python-path.diff
> index c400808085d..bb8b8a3ef60 100644
> --- a/devel/glib2-devel/files/patch-meson-build-python-path.diff
> +++ b/devel/glib2-devel/files/patch-meson-build-python-path.diff
> @@ -8,11 +8,11 @@ GLib2 tries to find "python3" and if it can't find it, it will go for "python";
>   glib_conf.set('HAVE_PROC_SELF_CMDLINE', have_proc_self_cmdline)
>   
>  -python = import('python').find_installation('python3')
> -+python = import('python').find_installation('')
> ++python = import('python').find_installation('@PYTHON@')
>   # used for '#!/usr/bin/env <name>'
>  -python_name = 'python3'
> -+python_name = '@PREFIX@/bin/python3.7'
> ++python_name = '@PYTHON@'
>   
>   python_version = python.language_version()
>   python_version_req = '>=3.4'
> - 
> \ No newline at end of file
> +
> ```

Yeah that's a better approach; I used a middle ground between that and what I had now. It'd be cool now if somebody could test the port on older systems.

Lord-Kamina commented:
So... is there anything particular still holding this behind?

mf2k commented:
The main issue is that it currently conflicts with the main branch.

Lord-Kamina commented:
@mf2k does it? 

<img width="810" alt="Screen Shot 2020-02-25 at 19 25 36" src="https://user-images.githubusercontent.com/6446892/75293326-b8194a00-5804-11ea-8ca7-15088b655c92.png">

mf2k commented:
Strange. Mine says "This branch cannot be rebased safely"

Lord-Kamina commented:
> Strange. Mine says "This branch cannot be rebased safely"

Weird... I'll try to rebase it on an updated master just to be sure *shrugs*

mf2k commented:
Unfortunately, I still see "This branch cannot be rebased safely". 

@pmetzger @cjones051073 What do you see?

emcrisostomo commented:
@mf2k , same thing here:

```
This branch cannot be rebased safely
```

emcrisostomo commented:
@Lord-Kamina: I see two commits. The first (2508fc5) is where the conflict is, while the second is fine. Actually, the second (4b6e1af) seems to be the interesting one, given the name.

If you aren't able to fix the problem, you could you try to:
* Branch from upstream/master.
* Cherry-picking the relevant commits.
* Squashing everything.
* Submit a new PR, and reference this one from a comment so that the conversation gets tracked

Lord-Kamina commented:
Which is really weird because my branch actually has 8 commits. So anyway, when I get some time I'll rebase the commits to streamline a bit and cherry-pick them onto a fresh branch or something

emcrisostomo commented:
Yes, but seven are here, where they appear as a merge:

https://github.com/macports/macports-ports/pull/5808/commits/2508fc5a0ebf2a4d2edaabe2cf8b80cfa3214ea9

Lord-Kamina commented:
> Yes, but seven are here, where they appear as a merge:
> 
> [2508fc5](https://github.com/macports/macports-ports/commit/2508fc5a0ebf2a4d2edaabe2cf8b80cfa3214ea9)

Weird. I originally did that merge to try and fix this issue. 🤷🏼‍♂️
Well, hopefully a clean rebase will get me out of the twilight zone.

Lord-Kamina commented:
How about now?

emcrisostomo commented on devel/glib2-devel/Portfile:
```diff
@@ -80,9 +81,10 @@ configure.python            ${prefix}/bin/python3.7
     reinplace -W ${worksrcpath} "s|@PREFIX@|${prefix}|g" gio/gdbusprivate.c gio/xdgmime/xdgmime.c glib/gi18n-lib.h glib/gi18n.h gio/gdbusaddress.c
     reinplace "s|data_dirs = \"/usr|data_dirs = \"${prefix}/share:/usr|g" ${worksrcpath}/glib/gutils.c
-    reinplace "s|path = \"/bin|path = \"${prefix}/bin:/bin|g" ${worksrcpath}/glib/gutils.c ${worksrcpath}/glib/gspawn.c
+    reinplace "s|path = \"/bin|path = \"${prefix}/bin:/bin|g" ${worksrcpath}/glib/gutils.c ${worksrcpath}/glib/gspawn.c    
```
There is a whitespace issue at the end of this line

Lord-Kamina replied:
Fixed.

emcrisostomo commented:
Much better. There is actually a small whitespace issue in the revision I've started. We're almost there.

emcrisostomo commented:
Added @ryandesign to the list of reviewers, to ask for his final opinion.

kencu commented:
I'm looking to kill off ancient PRs that aren't going anywhere -- is this one?

Lord-Kamina commented:
> I'm looking to kill off ancient PRs that aren't going anywhere -- is this one?

This PR is functional and eventually necessary (or one building up from it, at least); I've been using glib2 2.62 made from this PR and built several other libraries and apps depending on it successfully. I believe the main outstanding issue with the PR was the possible implications it could have for people on legacy systems.

kencu commented:
Well OK -- we will need to update glib2, no doubt about it. Is this a "push it and fix the fallout" situation?

Lord-Kamina commented:
> Well OK -- we will need to update glib2, no doubt about it. Is this a "push it and fix the fallout" situation?

In my opinion yes, but I think other people should weigh-in on it. The way I see it, one issue is @ryandesign is the person most qualified to see what the issues are but he's otherwise occupied, as he's said on a few different issues. From what he said in https://github.com/macports/macports-ports/pull/6544#issuecomment-598235781, it seems one of the blockers for updating to the meson versions of GNOME ports would be the universal variants. I have no way to test that, however.

kencu commented:
OK. So this PR would be OK it it could build universal, I guess. But it presently does/does not, nobody knows yet. At least that is an understandable blocker.

kencu commented:
I don't think anyone is going to touch the commit button on this until Ryan gets around to doing it.

kencu commented:
I need a newer glib2 now to build some software I'm working on.

I take it the issue is the powerpc build specifically, as I saw above, and I guess, building +universal on meson in general being kinda broken.

Let me take a look at this with you then.

kencu commented:
I fixed the issue with meson building universal. Perhaps that will let some of this old dead wood move along down the beach...

fredowski commented:
Hi, I need this updated glib2 also for the pspp quartz build.

kencu commented:
Ladies and Gentlemen -- WHATEVER THE H*LL we need to do to get glib2 updated, be it this PR or something else based off it, the time has long come and gone that we do it.

So let's do it.

kencu commented:
rebased.

The standard build succeeds (currently on BigSur Intel). The universal build fails. Log attached. Currently working on using the muniversal portgroup to fix that failure.

[glib2-devel-fail.txt](https://github.com/macports/macports-ports/files/5962866/glib2-devel-fail.txt)

kencu commented:
The universal build fails on BIgSur Intel even with the muniversal PG, sadly. Log attatched.

Here is the small patch needed to the Portfile to make meson even try to build universal with the muniversal PG:
```
% git diff
diff --git a/devel/glib2-devel/Portfile b/devel/glib2-devel/Portfile
index 4bb00193481..c009c9aba41 100644
--- a/devel/glib2-devel/Portfile
+++ b/devel/glib2-devel/Portfile
@@ -3,6 +3,7 @@
 PortSystem                  1.0
 PortGroup                   compiler_blacklist_versions 1.0
 PortGroup                   meson 1.0
+PortGroup                   muniversal 1.0
 
 # Please keep the glib2 and glib2-devel ports as similar as possible.
 
@@ -80,6 +81,17 @@ configure.perl              /usr/bin/perl
 configure.python            ${prefix}/bin/python3.9
 configure.env-append        PERL_PATH=${configure.perl}
 
+
+if {![info exists universal_possible]} {
+    set universal_possible [expr {${os.universal_supported} && [llength ${configure.universal_archs}] >= 2}]
+}
+if {${universal_possible} && [variant_isset universal]} {
+    foreach my_arch {ppc ppc64 i386 x86_64 arm64} {
+        # strip the automatic setting of host, meson does not accept
+        set merger_host(${my_arch}) ""
+    }
+}
+
 post-patch {
     reinplace -W ${worksrcpath} "s|@PYTHON@|${configure.python}|" meson.build
     reinplace -W ${worksrcpath} "s|@PREFIX@|${prefix}|g" gio/gdbusprivate.c gio/xdgmime/xdgmime.c glib/gi18n-lib.h glib/gi18n.h gio/gdbusaddress.c
```

kencu commented:
fixed that, and a failing test. 

Now I think just one more thing needs fixing -- the Portfile copies a header into the build, and when building universal, it needs to copy the header into all the build directories.

```
post-configure {
    system "ed - ${build.dir}/config.h < ${filespath}/config.h.ed"
}
```

kencu commented:
looks like once we get this up and running we should soon look towards an even newer version after that:
```
glib2 seems to have been updated (port version: 2.58.3, new version: 2.66.6
```

I would personally get this current one through glib2-devel and then into glib2, use that as a stepping stone, and once we do that make glib2-devel this newer version, rather than starting all over again from scratch.

kencu commented:
OK, that's got it, for all the systems I tried it on so far. 

I think next thing we do is push this and test out glib2-devel in the wild. We are not going to get much further doing it in this PR.
